ethereumj-core/src/main/java/org/ethereum/sync/SyncManager.java

```diff
@@ -155,6 +165,184 @@ public void init(final ChannelManager channelManager, final SyncPool pool) {
+                                        = channel.getEthHandler().sendGetBlockBodies(singletonList(blockHeaderWrapper));
+
+                                if (futureBlocks != null) {
+                                    // print needed to prevent an infinite loop when there are 2 blocks for download
+                                    logger.trace("futureBlocks number: " + futureBlocks.get().size());
+                                    Futures.addCallback(futureBlocks, new BlocksCallback(channel));
+                                }
+                            }
+                        }
+                    }
+
+                    int maxRequests = blocksToAsk / MAX_IN_REQUEST;
+                    int maxBlocks = MAX_IN_REQUEST * Math.min(maxRequests, REQUESTS);
+                    int reqBlocksCounter = 0;
+                    int blocksRequested = 0;
+                    Iterator<SyncQueueIfc.BlocksRequest> it = bReqs.iterator();
+                    while (it.hasNext() && blocksRequested < maxBlocks) {
+//                    for (SyncQueueIfc.BlocksRequest blocksRequest : bReq.split(MAX_IN_REQUEST)) {
+                        SyncQueueIfc.BlocksRequest blocksRequest = it.next();
+                        Channel any = getAnyPeer();
+                        if (any == null) {
+                            logger.debug("{} blockRetrieveLoop: No IDLE peers found", name);
+                            break;
+                        } else {
+                            logger.debug("{} blockRetrieveLoop: Requesting " + blocksRequest.getBlockHeaders().size() + " blocks from " + any.getNode(), name);
+                            ListenableFuture<List<Block>> futureBlocks
+                                    = any.getEthHandler().sendGetBlockBodies(blocksRequest.getBlockHeaders());
+                            blocksRequested += blocksRequest.getBlockHeaders().size();
+                            if (futureBlocks != null) {
+                                
+                                Futures.addCallback(futureBlocks, new BlocksCallback(any));
+                                reqBlocksCounter++;
+                                it.remove();
+                            }
+                        }
+                    }
+                    receivedBlocksLatch = new CountDownLatch(max(reqBlocksCounter - 2, 1));
+                    receivedBlocksLatch.await(1000, TimeUnit.MILLISECONDS);
+                } else {
+                    logger.debug("{} blockRetrieveLoop: BlockQueue is full", name);
+                    Thread.sleep(200);
+                }
+            } catch (InterruptedException e) {
+                break;
+            } catch (Exception e) {
+                logger.error("Unexpected: ", e);
+            }
+        }
+    }
+
+    @Override
+    protected void headerRetrieveLoop() {
+        List<SyncQueueIfc.HeadersRequest> hReq = emptyList();
+        while (!Thread.currentThread().isInterrupted() && !syncDone) {
```

Comment:
You shouldn't stop `headerRetrieveLoop` on syncDone, because it's not guaranteed that you will get all `NEW_BLOCK` messages without gaps